src/DiskIO/OIO/File.cc

```diff
@@ -0,0 +1,225 @@
+{
+    auto *aiocbp = static_cast<struct aiocb *>(lpOverlapped->hEvent);
+
+    aiocbp->aio_sigevent.sigev_notify = dwErrorCode;
+    aiocbp->aio_sigevent.sigev_signo = dwNumberOfBytesTransfered;
+    debugs(81, 7, "OIO operation complete: errorcode=" << dwErrorCode << " nbytes=" << dwNumberOfBytesTransfered);
+    xfree(lpOverlapped);
+}
+
+static HANDLE
+HandleFromFd(int fd)
+{
+    return reinterpret_cast<HANDLE>(_get_osfhandle(fd));
+}
+
+void
+DiskIO::OIO::File::read(ReadRequest *request)
+{
+    assert(strategy->aq.aq_state == AQ_STATE_SETUP);
+
+    /* Find a free slot */
+    const auto slot = strategy->findSlot();
+    if (slot < 0) {
+        /* No free slot? Callback error, and return */
+        debugs(79, DBG_IMPORTANT, "WARNING: out of aiocb slots!");
+        fatal("Aiee! out of aiocb slots! - TODO fix and wrap file_read\n");
+        /* fall back to blocking method */
+        //        file_read(fd, request->buf, request->len, request->offset, callback, data);
+        return;
+    }
+
+    /* Mark slot as ours */
+    auto qe = &strategy->aq.aq_queue[slot];
+    qe->aq_e_state = AQ_ENTRY_USED;
+    qe->aq_e_callback_data = cbdataReference(request);
+    qe->theFile = cbdataReference(this);
+    qe->aq_e_type = AQ_ENTRY_READ;
+    qe->aq_e_free = nullptr;
+    qe->aq_e_buf = request->buf;
+    qe->aq_e_fd = getFD();
+
+    /* Account */
+    ++ strategy->aq.aq_numpending;
+
+    /* Initiate I/O */
+    auto *aiocbp = &qe->aq_e_aiocb;
+    aiocbp->aio_fildes = getFD();
+    aiocbp->aio_nbytes = request->len;
+    aiocbp->aio_offset = request->offset;
+    aiocbp->aio_buf = request->buf;
+    aiocbp->aio_sigevent.sigev_notify = EINPROGRESS;
+    aiocbp->aio_sigevent.sigev_signo = -1;
+
+    auto Overlapped = (LPOVERLAPPED) xcalloc(1, sizeof(OVERLAPPED));
+#if _FILE_OFFSET_BITS==64
+#ifdef __GNUC__
+    Overlapped->Offset = (DWORD) (aiocbp->aio_offset % 0x100000000LL);
+    Overlapped->OffsetHigh = (DWORD) (aiocbp->aio_offset / 0x100000000LL);
+#else
+    Overlapped->Offset = (DWORD) (aiocbp->aio_offset % 0x100000000);
+    Overlapped->OffsetHigh = (DWORD) (aiocbp->aio_offset / 0x100000000);
+#endif
+#else
+    Overlapped->Offset = aiocbp->aio_offset;
+    Overlapped->OffsetHigh = 0;
+#endif
+    Overlapped->hEvent = aiocbp;
+
+    if (!ReadFileEx(HandleFromFd(aiocbp->aio_fildes), aiocbp->aio_buf, aiocbp->aio_nbytes, Overlapped, IoCompletionRoutine))
+        fatalf("Aiee! OIO ReadFileEx() returned error: %s\n", xstrerr(GetLastError()));
+}
+
+void
+DiskIO::OIO::File::write(WriteRequest *request)
+{
+    assert(strategy->aq.aq_state == AQ_STATE_SETUP);
+
+    /* Find a free slot */
+    const auto slot = strategy->findSlot();
+    if (slot < 0) {
+        /* No free slot? Callback error, and return */
+        debugs(79, DBG_IMPORTANT, "WARNING: out of oiocb slots!");
+        fatal("Aiee! out of oiocb slots TODO fix and wrap file_write !\n");
+        /* fall back to blocking method */
```

Comment:
I don't think we want to do this. I suspect we'd rather Assure() or something similar that terminates the current transaction but lets others continue

Reply:
Agreed. Though that will require run-time testing to ensure. For now I'm trying to leave the behaviour on Windows as unchanged as possible.